### DISCUSSION TITLE
能否针对 target 设置 option

### DISCUSSION THREAD

**Discussion by MasterBalll** [xmake-io/xmake#3388] [Q&A]
- 目前的需求是想要通过循环来定义若干 target，它们共用一套代码，只是配置不同（包括平台、工具链和宏的差别），目的是能够一键编译甚至运行全部。
- 现在尝试这么做，就是把那些差异化设置独立到几个 option 中，然后通过一个 table 来指定这些 option ，从而批量生成 target。
- 遇到几个问题，一个是 option 中不能设置工具链，这个倒是好说，放到 target 就好了，另外就是不太清楚怎么在 target 中设置 option ，set_config() 看起来是不能使用的。

xmake.lua 大概长这个样子。
```lua
local optsGen = {
    {sim = "hifi2,hifi3", ver = "1,2,3"}
}

local opts = {}
for _, optGen in ipairs(optsGen) do
    for _, simStr in ipairs(string.split(optGen.sim, ",")) do
        for _, verStr in ipairs(string.split(optGen.ver, ",")) do
            table.insert(opts, {sim = simStr , ver= verStr })
        end
    end
end

for _, opt in ipairs(opts) do

local algo = "ttt_" .. opt.sim .. "_" .. opt.mic or "ttt"

target(algo)
    set_kind("binary")
    -- set_config("ver", opt.ver)  ??
    add_options("sim", "ver")
    add_files("*.c")
target_end()

end

option("sim")
    set_default("pc")
option_end()

option("ver")
    set_default("1")
    if get_config("ver") == "1" then
        add_defines("ONE")
    elseif get_config("ver") == "2" then
        add_defines("TWO")
    end
option_end()
```

**Comment by waruqi** [comment]
> 遇到几个问题，一个是 option 中不能设置工具链，这个倒是好说，放到 target 就好了，另外就是不太清楚怎么在 target 中设置 option ，set_config() 看起来是不能使用的。

用法和思路不对，option 主要用于 选项配置，就是一个开关状态，尽可能打薄，你用反了。。

定义完基础 option 选项后，defines 这种条件配置，应该在 target 里面通过 `if is_config("ver", "1") then` 直接在 target 里面条件配置。

也可以在 on_load 里面，或者定义 rule 来更灵活配置，而不是塞到 option 里面做。。

> 目前的需求是想要通过循环来定义若干 target，它们共用一套代码，只是配置不同（包括平台、工具链和宏的差别），目的是能够一键编译甚至运行全部。

除了批量定义 target，也可以直接定义 task ，批量处理编译，只用一个 target

```
target("A")
    -- ...

task("build_all")
    on_run(function ()
        os.exec("xmake f --ver=1")
        os.exec("xmake")
        os.exec("xmake f --ver=2")
        os.exec("xmake")
        os.exec("xmake f --ver=3")
        os.exec("xmake")
    end)
```

然后只需要 

```
xmake build_all
```

就行了，而且 xmake 有宏批量记录插件，之后配置记录下操作命令，批量回放编译就行了，task 都不需要。

看下：https://xmake.io/#/zh-cn/plugin/builtin_plugins?id=%e5%ae%8f%e8%ae%b0%e5%bd%95%e5%92%8c%e5%9b%9e%e6%94%be

**Reply by MasterBalll** [reply]
> 用法和思路不对，option 主要用于 选项配置，就是一个开关状态，尽可能打薄，你用反了。。

原来 option 是这样的……
其实刚用 xmake 时就困惑于 option, rule 它们的侧重点是啥，随着使用才逐渐有点感觉，可能也是因为我没用过其他构建工具……

> 除了批量定义 target，也可以直接定义 task ，批量处理编译，只用一个 target

之前是写了 task 来着，但一方面是自定义的工具链有启动延时（[关联 issue](https://github.com/xmake-io/xmake/issues/3379#issue-1588720291)），在有几十种配置组合的情况下，每次随便改点东西，切配置等待检测都要等很久，实在很耗时；另一方面是依赖包有更新时要对每种配置回车安装一下，不能一下把所有配置的包都装好。
所以才想干脆改成每种配置对应一个 target 好了。

附上一个切配置等待的现象：

https://user-images.githubusercontent.com/122007764/219939779-fc352b44-b519-44e1-8804-fa763f9c0298.mp4

**Reply by waruqi** [reply]
这跟切配置无关，理论上不应该卡这么久，要么就是编译器自身有问题，运行会卡，你可以先切到单任务编译试试，或者禁用 cache 试试

```console
xmake f --ccache=n
xmake -j1
```

确认下是否 cache 或者并行编译导致，还是编译器自身存在问题。。

**Reply by MasterBalll** [reply]
关了 ccache ，做单任务构建，也是一样卡。
应该就是编译器、链接器自身的问题。这套工具链运行时要鉴权，所以每检测一次就要等一下。
不知道咋办能加速，安装同一个包的多种配置写起来也好麻烦，暂时先用回单 target ，定义 task 的方式了。
另外问下，是否有选项能跳过包安装的确认步骤，直接自动安装？

**Reply by star-hengxing** [reply]
> 另外问下，是否有选项能跳过包安装的确认步骤，直接自动安装？

```
xmake -y
```

**Reply by MasterBalll** [reply]
眼拙了……原来是这个
